akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala

```diff
@@ -452,6 +453,66 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
+        sendWINDOW_UPDATE(0, missingWindow)
+        sendWINDOW_UPDATE(TheStreamId, missingWindow)
+
+        val entityDataOut = TestPublisher.probe[ByteString]()
+
+        val response = HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, Source.fromPublisher(entityDataOut)))
+        emitResponse(TheStreamId, response)
+        expectDecodedResponseHEADERS(streamId = TheStreamId, endStream = false) shouldBe response.withEntity(HttpEntity.Empty.withContentType(ContentTypes.`application/octet-stream`))
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+
+        expectDATA(TheStreamId, false, bytesToSend)
+
+        entityDataOut.sendComplete()
+        expectDATA(TheStreamId, true, 0)
+      }
+
+      "handle WINDOW_UPDATE correctly when received before started sending out response" in new WaitingForResponseSetup {
+        val bytesToSend = 70000 // > Http2Protocol.InitialWindowSize
+        val missingWindow = bytesToSend - Http2Protocol.InitialWindowSize
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow)
+        sendWINDOW_UPDATE(TheStreamId, missingWindow)
+
+        val entityDataOut = TestPublisher.probe[ByteString]()
+
+        val response = HttpResponse(entity = HttpEntity(ContentTypes.`application/octet-stream`, Source.fromPublisher(entityDataOut)))
+        emitResponse(TheStreamId, response)
+        expectDecodedResponseHEADERS(streamId = TheStreamId, endStream = false) shouldBe response.withEntity(HttpEntity.Empty.withContentType(ContentTypes.`application/octet-stream`))
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+
+        expectDATA(TheStreamId, false, bytesToSend)
+
+        entityDataOut.sendComplete()
+        expectDATA(TheStreamId, true, 0)
+      }
+
+      "distribute increases to SETTINGS_INITIAL_WINDOW_SIZE to streams correctly while sending out response" in new WaitingForResponseDataSetup {
+        // changes to SETTINGS_INITIAL_WINDOW_SIZE need to be distributed to active streams: https://httpwg.org/specs/rfc7540.html#InitialWindowSize
+        val bytesToSend = 70000 // > Http2Protocol.InitialWindowSize
+        val missingWindow = bytesToSend - Http2Protocol.InitialWindowSize
+        require(missingWindow >= 0)
+        sendWINDOW_UPDATE(0, missingWindow) // SETTINGS_INITIAL_WINDOW_SIZE only has effect on stream-level window, so need to make room in connection
+
+        entityDataOut.sendNext(bytes(bytesToSend, 0x23))
+        expectDATA(TheStreamId, false, Http2Protocol.InitialWindowSize)
+        expectNoBytes(100.millis)
+
+        sendSETTING(Http2Protocol.SettingIdentifier.SETTINGS_INITIAL_WINDOW_SIZE, bytesToSend)
+        updateFromServerWindows(TheStreamId, _ + missingWindow) // test probe doesn't automatically update window
+        expectDATA(TheStreamId, false, missingWindow)
+        expectSettingsAck() // FIXME: bug: we must send ACK before making use of the new setting, see https://github.com/akka/akka-http/issues/3553
```

Comment:
@ignasi35 this test found the issue around applying settings we discussed in #3553.

Reply:
So when this bug is fixed we should see the Ack before we see the DATA? Make sense.

Reply:
I'm not sure I'm familiar enought with the issue. In cases when the window is exhausted, no frames should be produced and buffered until (1) a `WINDOW_UPDATE` is received, or (2) a `SETTINGS` is received and the corresponding `SETTING_ACK` is sent out.

When #3553 is fixed, I agree that `SETTINGS_ACK` should arrive before the SETTINGS ACK.

The problem is, when the server window on the client side is exhausted, the client should not pre-render any frames until a WINDOW_UPDATE (or a SETTINGS_INITIAL_WINDOW_SIZE) is received since there are no guarantees a pre-rendered frame would fit on the window, right? The same happens with SETTINGS_FRAME_SIZE.

------

Hmm, increasing the window size via settings is very different from the processing of a `WINDOW_UPDATE`. 

In `WINDOW_UPDATE`, the sender increases the window value and then sends the frame out so the remote peer's value for the window is always less or equal to the sender's value.
Using `SETTINGS_INITIAL_WINDOW_SIZE` means the sender doesn't increase the window value until it gets an `SETTINGS_ACK` but the remote peer already uses the increased window value so the remote peer's value for the window may be greater than the sender's window value.

It sounds like `SETTINGS_INITIAL_WINDOW_SIZE` as a broadcast replacement of `WINDOW_UPDATE` is a hack.

Reply:
> The problem is, when the server window on the client side is exhausted, the client should not pre-render any frames until a WINDOW_UPDATE (or a SETTINGS_INITIAL_WINDOW_SIZE) is received since there are no guarantees a pre-rendered frame would fit on the window, right? The same happens with SETTINGS_FRAME_SIZE.

Yes.

> In `WINDOW_UPDATE`, the sender increases the window value and then sends the frame out so the remote peer's value for the window is always less or equal to the sender's value.
> Using `SETTINGS_INITIAL_WINDOW_SIZE` means the sender doesn't increase the window value until it gets an `SETTINGS_ACK` but the remote peer already uses the increased window value so the remote peer's value for the window may be greater than the sender's window value.

For the right value of sender (i.e. the one that proposed the setting), that seems correct as well.

> It sounds like `SETTINGS_INITIAL_WINDOW_SIZE` as a broadcast replacement of `WINDOW_UPDATE` is a hack.

Agreed, that's not what it is supposed to be used for. I guess it was really originally thought to be used for setting the initial window size for new streams. Maybe it was changed this way because the sender can just ignore the SETTINGS ACK and assume the setting to be applied instantly and retroactively, so that it doesn't matter for the client if a received stream was created slightly before the corresponding ACK or slightly after.